lib/makara/notifications.rb

```diff
@@ -0,0 +1,40 @@
+module Makara
+  # A notification handler that allows users to register arbitrary instrumentation callbacks. For instance, this can be
+  # used to fire dtrace or strace probes, or to write log messages. Callbacks are in the form of blocks, which can take
+  # 0 to N arguments, where N is the number of arguments provided to the the :notify!: method.
+  #
+  # Example:
+  #
+  #   Makara::Cache instruments the :write: method as follows:
+  #
+  #     Notifications.notify!('Cache:write', key, value, ttl)
```

Comment:
With dtrace, USDT probes need to be registered with specific argument types. For instance, this is how we have been doing things: https://github.com/wanelo/makara/blob/dtrace/lib/makara/dtrace_provider.rb. It calls through to C bindings, so if that is not done correctly the process segfaults. *After* probes are registered, you can subscribe to them on the command line to make them actually emit information, so that's where I would listen to the `ruby*:makara::`

DTrace is fairly low-level, and you generally need to know how many arguments there are as well as the argument type. I'm not sure if other dynamic tracing libraries are friendlier on that front.